File: 2024_cg/hw5/load_mesh.py
import os
import numpy as np

class Vector3:
    def __init__(self, x=0, y=0, z=0):
        self.x = x
        self.y = y
        self.z = z

class Triangle:
    def __init__(self, indices):
        self.indices = indices

gPositions = []
gNormals = []
gTriangles = []

def tokenize(string, delimiter):
    tokens = string.split(delimiter)
    return tokens

def face_index(string):
    tokens = tokenize(string, "/")
    if tokens[0] and tokens[-1] and int(tokens[0]) == int(tokens[-1]):
        return int(tokens[0])
    else:
        print("ERROR: Bad face specifier!")
        exit(0)

def load_mesh(fileName):
    if not os.path.isfile(fileName):
        print(f"ERROR: Unable to load mesh from {fileName}!")
        exit(0)

    xmin, ymin, zmin = float('inf'), float('inf'), float('inf')
    xmax, ymax, zmax = float('-inf'), float('-inf'), float('-inf')

    with open(fileName, 'r') as fin:
        for line in fin:
            if line.startswith('v '):
                tokens = tokenize(line[2:], ' ')
                x, y, z = float(tokens[0]), float(tokens[1]), float(tokens[2])
                gPositions.append(Vector3(x, y, z))
                xmin, ymin, zmin = min(xmin, x), min(ymin, y), min(zmin, z)
                xmax, ymax, zmax = max(xmax, x), max(ymax, y), max(zmax, z)
            elif line.startswith('vn '):
                tokens = tokenize(line[3:], ' ')
                x, y, z = float(tokens[0]), float(tokens[1]), float(tokens[2])
                gNormals.append(Vector3(x, y, z))
            elif line.startswith('f '):
                tokens = tokenize(line[2:], ' ')
                indices = [face_index(t) - 1 for t in tokens[:3]]  # Subtracting 1 to convert to 0-based index
                gTriangles.append(Triangle(indices))

    print(f"Mesh loaded with {len(gPositions)} vertices, {len(gNormals)} normals, and {len(gTriangles)} triangles.")
    print(f"Bounding box: xmin={xmin}, ymin={ymin}, zmin={zmin}, xmax={xmax}, ymax={ymax}, zmax={zmax}")

# Example usage
load_mesh("bunny.obj")

File: 2024_cg/hw5/cpp/load_mesh.cpp
/*
	How to use this code:

	Call load_mesh("bunny.obj") after creating the window using GLUT. This will
	read the mesh data from bunny.obj and populate the arrays gPositions,
	gNormals, and gTriangles.

	When rendering, we use a similar convention to the sphere from PA2. In other
	words, for triangle i, define:

		k0 = gTriangles[i].indices[0];
		k1 = gTriangles[i].indices[1];
		k2 = gTriangles[i].indices[2];

	Then the vertices of the triangle are at gPositions[k0], gPositions[k1], and
	gPositions[k2], in that order. The normals of the corresponding vertices are
	at gNormals[k0], gNormals[k1], and gNormals[k2], in that order.

	For the second part of the assignment, you have to copy the gPositions,
	gNormals, and gTriangles arrays into GPU memory.
*/

#include <stdio.h>
#include <string.h>
#include <string>
#include <vector>
#include <queue>
#include <fstream>
#include <float.h>

struct Vector3
{
	float			x, y, z;
};

struct Triangle
{
	unsigned int 	indices[3];
};

std::vector<Vector3>	gPositions;
std::vector<Vector3>	gNormals;
std::vector<Triangle>	gTriangles;

void tokenize(char* string, std::vector<std::string>& tokens, const char* delimiter)
{
	char* token = strtok(string, delimiter);
	while (token != NULL)
	{
		tokens.push_back(std::string(token));
		token = strtok(NULL, delimiter);
	}
}

int face_index(const char* string)
{
	int length = strlen(string);
	char* copy = new char[length + 1];
	memset(copy, 0, length+1);
	strcpy(copy, string);

	std::vector<std::string> tokens;
	tokenize(copy, tokens, "/");
	delete[] copy;
	if (tokens.front().length() > 0 && tokens.back().length() > 0 && atoi(tokens.front().c_str()) == atoi(tokens.back().c_str()))
	{
		return atoi(tokens.front().c_str());
	}
	else
	{
		printf("ERROR: Bad face specifier!\n");
		exit(0);
	}
}

void load_mesh(std::string fileName)
{
	std::ifstream fin(fileName.c_str());
	if (!fin.is_open())
	{
		printf("ERROR: Unable to load mesh from %s!\n", fileName.c_str());
		exit(0);
	}

	float xmin = FLT_MAX;
	float xmax = -FLT_MAX;
	float ymin = FLT_MAX;
	float ymax = -FLT_MAX;
	float zmin = FLT_MAX;
	float zmax = -FLT_MAX;

	while (true)
	{
		char line[1024] = {0};
		fin.getline(line, 1024);

		if (fin.eof())
			break;

		if (strlen(line) <= 1)
			continue;

		std::vector<std::string> tokens;
		tokenize(line, tokens, " ");

		if (tokens[0] == "v")
		{
			float x = atof(tokens[1].c_str());
			float y = atof(tokens[2].c_str());
			float z = atof(tokens[3].c_str());

			xmin = std::min(x, xmin);
			xmax = std::max(x, xmax);
			ymin = std::min(y, ymin);
			ymax = std::max(y, ymax);
			zmin = std::min(z, zmin);
			zmax = std::max(z, zmax);

			Vector3 position = {x, y, z};
			gPositions.push_back(position);
		}
		else if (tokens[0] == "vn")
		{
			float x = atof(tokens[1].c_str());
			float y = atof(tokens[2].c_str());
			float z = atof(tokens[3].c_str());
			Vector3 normal = {x, y, z};
			gNormals.push_back(normal);
		}
		else if (tokens[0] == "f")
		{
			unsigned int a = face_index(tokens[1].c_str());
			unsigned int b = face_index(tokens[2].c_str());
			unsigned int c = face_index(tokens[3].c_str());
			Triangle triangle;
			triangle.indices[0] = a - 1;
			triangle.indices[1] = b - 1;
			triangle.indices[2] = c - 1;
			gTriangles.push_back(triangle);
		}
	}

	fin.close();

	printf("Loaded mesh from %s. (%lu vertices, %lu normals, %lu triangles)\n", fileName.c_str(), gPositions.size(), gNormals.size(), gTriangles.size());
	printf("Mesh bounding box is: (%0.4f, %0.4f, %0.4f) to (%0.4f, %0.4f, %0.4f)\n", xmin, ymin, zmin, xmax, ymax, zmax);
}
